tests/func/artifacts/test_artifacts.py

```diff
@@ -0,0 +1,131 @@
+import os
+
+import pytest
+
+from dvc.annotations import Artifact
+from dvc.repo.artifacts import name_is_compatible
+from dvc.utils.strictyaml import YAMLSyntaxError, YAMLValidationError
+
+dvcyaml = {
+    "artifacts": {
+        "myart": {"type": "model"},
+        "hello": {"type": "file", "path": "hello.txt"},
+        "world": {
+            "type": "object",
+            "path": "world.txt",
+            "desc": "The world is not enough",
+            "labels": ["but", "this", "is"],
+            "meta": {"such": "a", "perfect": "place to start"},
+        },
+    }
+}
+
+
+def set_id_as_path(artifacts):
+    artifacts = artifacts.copy()
+    for name in artifacts:
+        artifacts[name]["path"] = artifacts[name].get("path", name)
+    return artifacts
+
+
+def test_reading_artifacts_subdir(tmp_dir, dvc):
+    (tmp_dir / "dvc.yaml").dump(dvcyaml)
+
+    subdir = tmp_dir / "subdir"
+    subdir.mkdir()
+
+    (subdir / "dvc.yaml").dump(dvcyaml)
+
+    artifacts = {
+        name: Artifact(**values)
+        for name, values in set_id_as_path(dvcyaml["artifacts"]).items()
+    }
+    assert tmp_dir.dvc.artifacts.read() == {
+        "dvc.yaml": artifacts,
+        f"subdir{os.path.sep}dvc.yaml": artifacts,
+    }
+
+
+bad_dvcyaml_extra_field = {
+    "artifacts": {
+        "lol": {"kek": "cheburek"},
+        "hello": {"type": "file", "path": "hello.txt"},
+    }
+}
+
+
+def test_broken_dvcyaml_extra_field(tmp_dir, dvc):
+    (tmp_dir / "dvc.yaml").dump(bad_dvcyaml_extra_field)
+
+    with pytest.raises(YAMLValidationError):
+        tmp_dir.dvc.artifacts.read()
+
+
+bad_dvcyaml_id_duplication = """
+artifacts:
+  lol:
+    type: kek
+  lol: {}
+"""
+
+
+def test_broken_dvcyaml_id_duplication(tmp_dir, dvc):
+    with open(tmp_dir / "dvc.yaml", "w") as f:
+        f.write(bad_dvcyaml_id_duplication)
+
+    with pytest.raises(YAMLSyntaxError):
+        tmp_dir.dvc.artifacts.read()
+
+
+dvcyaml_redirecting = {"artifacts": "artifacts.yaml"}
+
+
+def test_read_artifacts_yaml(tmp_dir, dvc):
+    (tmp_dir / "dvc.yaml").dump(dvcyaml_redirecting)
+    (tmp_dir / "artifacts.yaml").dump(dvcyaml["artifacts"])
+
+    artifacts = {
+        name: Artifact(**values)
+        for name, values in set_id_as_path(dvcyaml["artifacts"]).items()
+    }
+    assert tmp_dir.dvc.artifacts.read() == {
+        "dvc.yaml": artifacts,
+    }
+
+
+@pytest.mark.parametrize(
+    "name",
+    [
+        "nn",
+        "m1",
+        "model-prod",
+        "model-prod-v1",
+        "namespace/model",
+    ],
+)
+def test_check_name_is_valid(name):
+    assert name_is_compatible(name)
+
+
+@pytest.mark.parametrize(
+    "name",
+    [
+        "",
+        "m",
```

Comment:
Isn't this a valid tag name? Why is it disallowed?

Reply:
Actually a lot of these are allowed. I can see the regex above, but what's the logical reason for this particular set of rules and how do we explain them?

Reply:
Yep, this is a valid tag name.
1. This regex was set to be simple and have some constraints, so we could lift them up later (easier to do that to introduce new constraints). We use some special symbols like `@`, `#`, `!` and not all of them were introduced to Git tag right away. Maybe we'll need something else and encountering them in already existing tags will create ambiguity.
2. In DVC it's a recommendation - because DVC won't fail, only issue a warning. If you don't need Git tags (not sure whether this is a solid use case though), you can ignore this.
3. In GTO we have these limitation now (as per regex), so a Git tag `m@v0.0.1` won't be interpreted as something that registers artifact `m` of version `v0.0.1`.
4. We also should keep in mind that these ID should be user-facing (in Studio MR) and those users are potentially  managers or non-engineer folks.

So two solutions:
1. Keep as is
2. Try to lift some limitations - in both GTO and DVC

I think we should take 1st for now. I can keep this on the list, and later we can lift some constrains if we'll be getting requests and if we'll see it's ok.

Re how this can be communicated: adding it to docs, I've added it to dvcyaml-schema already, and issuing a warning in CLI when accessing artifacts may be enough.

WDYT?

Reply:
ah, and in fact the current regex doesn't support paths! Missed that at first. So we need to expand it then, at least to bare minimum that will support using path as ID.

Reply:
Here's a good docs that we can point to: https://docs.github.com/en/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names